### PR TITLE
First pass MVP chunked file upload

### PR DESCRIPTION
**Pull request by caseychoiniere**

### CODE REVIEW

File: app/scripts/stores/projectStore.js [hunk 1]
```diff
@@ -484,6 +525,76 @@ var ProjectStore = Reflux.createStore({
...
+                    break;
+                }
+            }
+        }
+
+        for (let i = 0; i < chunks.length; i++) {
+            let chunk = chunks[i];
+            if (chunk.status === this.STATUS_WAITING_FOR_UPLOAD || chunk.status === this.STATUS_RETRY) {
+                chunk.status = this.STATUS_UPLOADING;
+                ProjectActions.getChunkUrl(uploadId, upload.blob.slice(chunk.start, chunk.end), chunk.number, upload.size, upload.parentId, upload.parentKind);
```
Comment (jonturneratduke) [inline app/scripts/stores/projectStore.js:575]: if getChunkUrl fails (ex: dodgy network), this leaves chunk status as uploading, but won't retry and unable to finish. any way to trap around getChunkUrl?

File: app/scripts/stores/projectStore.js [hunk 2]
```diff
@@ -446,9 +487,9 @@ var ProjectStore = Reflux.createStore({
         let host = json.host;
         let url = json.url;
         var win = window.open(host + url, '_blank');
-        if(win){
+        if (win) {
             win.focus();
-        }else{
+        } else {
             alert('Please allow popups for this site and try downloading again');
         }
```
Comment (jonturneratduke) [inline app/scripts/stores/projectStore.js:494]: Nice catch! This is going to affect a lot of our users. You just saved a ton of helpdesk calls and tickets.

File: app/scripts/stores/projectStore.js [hunk 3]
```diff
@@ -5,6 +5,15 @@ import cookie from 'react-cookie';
...
 
     init() {
+        // Status ENUM
+        this.STATUS_WAITING_FOR_UPLOAD = null;
+        this.STATUS_SUCCESS = 1;
+        this.STATUS_RETRY = 2;
+        this.STATUS_UPLOADING = 3;
+        this.STATUS_FAILED = 4;
+
+        this.MAX_RETRY = 2;
```
Comment (jonturneratduke) [inline app/scripts/stores/projectStore.js:15]: Candidate for "performance tuning" once we have some real-world tests.

File: app/scripts/actions/projectActions.js [hunk 4]
```diff
@@ -440,6 +451,193 @@ ProjectActions.getDownloadUrl.preEmit = function (id) {
     })
 };
 
+ProjectActions.startUpload.preEmit = function (projId, blob, parentId, parentKind) {
+    let chunkNum = 0,
+        fileName = blob.name,
+        contentType = blob.type,
+        BYTES_PER_CHUNK, SIZE, NUM_CHUNKS, start, end;
+    BYTES_PER_CHUNK = 2097152;
```
Comment (jonturneratduke) [inline app/scripts/actions/projectActions.js:459]: Candidate for "performance tuning" later...

File: app/scripts/stores/projectStore.js [hunk 5]
```diff
@@ -484,6 +525,76 @@ var ProjectStore = Reflux.createStore({
...
+    startUpload() {
+        this.trigger({
+            uploading: true
+        })
+    },
+
+    startUploadSuccess(uploadId, details) {
+        this.uploads[uploadId] = details;
+        ProjectActions.updateAndProcessChunks(uploadId, null, null);
+    },
```
Comment (caseychoiniere) [inline app/scripts/stores/projectStore.js:539]: Add trigger change to strore

File: app/scripts/actions/projectActions.js [hunk 6]
```diff
@@ -440,6 +451,193 @@ ProjectActions.getDownloadUrl.preEmit = function (id) {
...
+
+        }).then(checkResponse).then(function (response) {
+            return response.json()
+        }).then(function (json) {
+            let chunkObj = json;
+            if (chunkObj && chunkObj.url && chunkObj.host) {
+                // upload chunks
+                uploadChunk(uploadId, chunkObj.host + chunkObj.url, chunkBlob, size, parentId, parentKind, chunkNum)
+            } else {
+                console.log("Problem, no upload created");
```
Comment (caseychoiniere) [inline app/scripts/actions/projectActions.js:549]: Change this

File: app/scripts/actions/projectActions.js [hunk 7]
```diff
@@ -440,6 +451,193 @@ ProjectActions.getDownloadUrl.preEmit = function (id) {
...
+            }
+            ProjectActions.startUploadSuccess()
+        }).catch(function (ex) {
+            ProjectActions.startUploadError(ex)
```
Comment (dburdick) [inline app/scripts/actions/projectActions.js:553]: call updateAndProcessChunks with uploadId, chunkNum and status of RETRY

File: app/scripts/actions/projectActions.js [hunk 8]
```diff
@@ -440,6 +451,193 @@ ProjectActions.getDownloadUrl.preEmit = function (id) {
...
+            return response.json()
+        }).then(function (json) {
+            let chunkObj = json;
+            if (chunkObj && chunkObj.url && chunkObj.host) {
+                // upload chunks
+                uploadChunk(uploadId, chunkObj.host + chunkObj.url, chunkBlob, size, parentId, parentKind, chunkNum)
+            } else {
+                console.log("Problem, no upload created");
+            }
+            ProjectActions.startUploadSuccess()
```
Comment (dburdick) [inline app/scripts/actions/projectActions.js:551]: remove this

File: app/scripts/actions/projectActions.js (hunk 6 above)
Comment (dburdick) [inline app/scripts/actions/projectActions.js:549]: here raise an exception saying upload not created

File: app/scripts/stores/projectStore.js [hunk 9]
```diff
@@ -484,6 +525,76 @@ var ProjectStore = Reflux.createStore({
...
+
+        for (let i = 0; i < chunks.length; i++) {
+            let chunk = chunks[i];
+            if (chunk.status === this.STATUS_WAITING_FOR_UPLOAD || chunk.status === this.STATUS_RETRY) {
+                chunk.status = this.STATUS_UPLOADING;
+                ProjectActions.getChunkUrl(uploadId, upload.blob.slice(chunk.start, chunk.end), chunk.number, upload.size, upload.parentId, upload.parentKind);
+                return;
+            }
+        }
+        ProjectActions.allChunksUploaded(uploadId, upload.parentId, upload.parentKind);
```
Comment (dburdick) [inline app/scripts/stores/projectStore.js:579]: only call if allDone is true

File: app/scripts/stores/projectStore.js [hunk 10]
```diff
@@ -484,6 +525,76 @@ var ProjectStore = Reflux.createStore({
...
+            }
+        }
+
+        for (let i = 0; i < chunks.length; i++) {
+            let chunk = chunks[i];
+            if (chunk.status === this.STATUS_WAITING_FOR_UPLOAD || chunk.status === this.STATUS_RETRY) {
+                chunk.status = this.STATUS_UPLOADING;
+                ProjectActions.getChunkUrl(uploadId, upload.blob.slice(chunk.start, chunk.end), chunk.number, upload.size, upload.parentId, upload.parentKind);
+                return;
+            }
```
Comment (dburdick) [inline app/scripts/stores/projectStore.js:577]: add check for `chunk.status === this.STATUS_UPLOADING` and set `allDone = false` if so, and continue looping

File: app/scripts/actions/projectActions.js [hunk 11]
```diff
@@ -440,6 +451,193 @@ ProjectActions.getDownloadUrl.preEmit = function (id) {
...
+            'Authorization': appConfig.apiToken,
+            'Accept': 'application/json'
+        }
+    }).then(checkResponse).then(function (response) {
+        return response.json()
+    }).then(function (json) {
+        ProjectActions.addFile(uploadId, parentId, parentKind);
+        console.log('File Upload Complete');
+    }).catch(function (ex) {
+        console.log('File Upload Failed');
```
Comment (caseychoiniere) [inline app/scripts/actions/projectActions.js:608]: upload failed action

File: app/scripts/stores/projectStore.js [hunk 12]
```diff
@@ -249,13 +260,43 @@ var ProjectStore = Reflux.createStore({
...
+        if (parentKind === 'dds-project') {
+            ProjectActions.loadProjectChildren(id);
+        } else {
+            ProjectActions.loadFolderChildren(id);
+        }
+        if (this.uploads.hasOwnProperty(uploadId)) {
+            delete this.uploads[uploadId];
+        }
+        this.trigger({
+            uploading: false,
```
Comment (caseychoiniere) [inline app/scripts/stores/projectStore.js:286]: take out uploading and base code on uploads

File: app/scripts/components/projectComponents/projectChildren.jsx [hunk 13]
```diff
@@ -15,6 +16,8 @@ var mui = require('material-ui'),
 class ProjectChildren extends React.Component {
 
     render() {
+        let uploading = this.props.uploading ? <div><LinearProgress color={"#2196f3"} mode="determinate" value={ProjectStore.uploadProgress} style={styles.uploader}/><div className="mdl-color-text--grey-600" style={styles.uploadText}>uploading...</div></div> : '';
```
Comment (dburdick) [inline app/scripts/components/projectComponents/projectChildren.jsx:19]: define uploading as a map over the keys of this.props.uploads. return a LinearProgress for each upload in uploads. include filename in text, i.e. "Uploading file.txt"

File: app/scripts/actions/projectActions.js [hunk 14]
```diff
@@ -440,6 +451,193 @@ ProjectActions.getDownloadUrl.preEmit = function (id) {
...
+        }).then(checkResponse).then(function (response) {
+            return response.json()
+        }).then(function (json) {
+            let chunkObj = json;
+            if (chunkObj && chunkObj.url && chunkObj.host) {
+                // upload chunks
+                uploadChunk(uploadId, chunkObj.host + chunkObj.url, chunkBlob, size, parentId, parentKind, chunkNum)
+            } else {
+                console.log("Problem, no upload created");
+            }
```
Comment (caseychoiniere) [inline app/scripts/actions/projectActions.js]: sdfsdf

**Pull request merged**